### PR TITLE
diesel pools launch revisions

### PR DESCRIPTION
**Pull request by donchate**
- restricts `sharesOut` and `maxSlippage` precision to <= 3
- fix error messages to show actual amount being compared in `validateSwap`
- change to string asserts on user input
- **change `addLiquidity` validation to calculate optimal values within new tolerance parameter**

Should not modify schema or any existing data in the DB.

### CODE REVIEW

File: contracts/marketpools.js [hunk 1]
```diff
@@ -181,14 +180,22 @@ actions.addLiquidity = async (payload) => {
...
-    || !api.assert(baseQuantity && api.BigNumber(baseQuantity).gt(0), 'invalid baseQuantity')
-    || !api.assert(quoteQuantity && api.BigNumber(quoteQuantity).gt(0), 'invalid quoteQuantity')
+    || !api.assert(typeof baseQuantity === 'string' && api.BigNumber(baseQuantity).gt(0), 'invalid baseQuantity')
+    || !api.assert(typeof quoteQuantity === 'string' && api.BigNumber(quoteQuantity).gt(0), 'invalid quoteQuantity')
     || !await validateTokenPair(tokenPair)) return;
 
+  let addSlippage = api.BigNumber('0.01');
+  if (maxSlippage) {
+    if (!api.assert(typeof maxSlippage === 'string'
+      && api.BigNumber(maxSlippage).gt(0) && api.BigNumber(maxSlippage).lt(50), 'maxSlippage must be greater than 0 and less than 50')) return;
```
Comment (eonwarped) [inline contracts/marketpools.js:195]: Do we want to allow 50?

Comment (CodeBull) [reply contracts/marketpools.js:195]: Without high slippage pools with low liquidity can't be used. So, may be allowing it in the contract is fine, users will make informed decision to change it to whatever % they like.

Comment (donchate) [reply contracts/marketpools.js:195]: All that will happen is they'll potentially deposit 50% less than they intended into the pool.

**Review by bt-cryptomancer — CHANGES_REQUESTED** [review]
Can you add validation for a maximum allowed precision on maxSlippage in both the swapTokens and addLiquidity actions?  I don't think we should ever allow arbitrary precision user input, it should be limited to some reasonable number of digits.

Also, I'm not sure I agree with restricting sharesOut to be an integer only.  I can imagine some cases where there's a pool with very large liquidity, when an integer sharesOut might not offer a fine grained enough demarcation.  Perhaps something like 3 or 4 decimal digits of precision would be better.

**Comment by donchate** [comment]
Applied precision <= 3 for maxSlippage, sharesOut

File: contracts/marketpools.js [hunk 2]
```diff
@@ -199,21 +206,42 @@ actions.addLiquidity = async (payload) => {
...
+        if (!api.assert(api.BigNumber(quoteOptimal).gte(quoteMin), 'exceeded max slippage for adding liquidity')) return;
+        amountAdjusted = [baseQuantity, quoteOptimal];
+      } else {
+        const baseOptimal = getQuote(quoteQuantity, pool.quoteQuantity, pool.baseQuantity).toFixed(baseToken.precision, api.BigNumber.ROUND_HALF_UP);
+        if (api.BigNumber(baseOptimal).lte(baseQuantity)) {
+          if (!api.assert(api.BigNumber(baseOptimal).gte(baseMin), 'exceeded max slippage for adding liquidity')) return;
+          amountAdjusted = [baseOptimal, quoteQuantity];
+        }
+      }
+      const pPrecision = Math.min(baseToken.precision, quoteToken.precision);
```
Comment (eonwarped) [inline contracts/marketpools.js:225]: Want to re-raise this, but in terms of precision, should it not be by the token unit of this price (either quote or base, not sure which)

**Comment by donchate** [comment]
Pending changes to resolve newly discovered test case.

**Comment by donchate** [comment]
The new optimal values on addLiquidity are calculated using verified math to preserve price. The outputs are also verified in test coverage. The additional validation was failing (incorrectly) due to two stages of rounding and it appears it can be removed entirely.

**Review by bt-cryptomancer — APPROVED** [review]
Changes look good.

**Pull request merged**